crates/wit-parser/src/sizealign.rs

```diff
@@ -1,26 +1,210 @@
+use std::{
+    num::NonZeroUsize,
+    ops::{Add, AddAssign},
+};
+
 use crate::{FlagsRepr, Int, Resolve, Type, TypeDef, TypeDefKind};
 
-#[derive(Default)]
-pub enum AddressSize {
-    #[default]
-    Wasm32,
-    Wasm64,
+/// Architecture specific alignment
+#[derive(Eq, PartialEq, PartialOrd, Clone, Copy, Debug)]
+pub enum Alignment {
+    Pointer,
+    Bytes(NonZeroUsize),
 }
 
-#[derive(Default)]
-pub struct SizeAlign {
-    map: Vec<(usize, usize)>,
-    wasm_type: AddressSize,
+impl Default for Alignment {
+    fn default() -> Self {
+        Alignment::Bytes(NonZeroUsize::new(1).unwrap())
+    }
+}
+
+impl std::fmt::Display for Alignment {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Alignment::Pointer => f.write_str("ptr"),
+            Alignment::Bytes(b) => f.write_fmt(format_args!("{}", b.get())),
+        }
+    }
+}
+
+impl Ord for Alignment {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match (self, other) {
+            (Alignment::Pointer, Alignment::Pointer) => std::cmp::Ordering::Equal,
+            (Alignment::Pointer, Alignment::Bytes(b)) => {
+                if b.get() > 4 {
+                    std::cmp::Ordering::Less
+                } else {
+                    std::cmp::Ordering::Greater
+                }
+            }
+            (Alignment::Bytes(b), Alignment::Pointer) => {
+                if b.get() > 4 {
+                    std::cmp::Ordering::Greater
+                } else {
+                    std::cmp::Ordering::Less
+                }
+            }
+            (Alignment::Bytes(a), Alignment::Bytes(b)) => a.cmp(b),
+        }
+    }
+}
```

Comment:
Could you have some code comments for what ordering this is achieving? It's not immediately clear to this myself but I think some comments would help out indicating what order is being achieved.

Reply:
The answer is easy to state (I will add some comments):

 Bytes(1) < Bytes(2) < Bytes(4) <= Pointer <= Bytes(8)
 
 This is needed to calculate max alignment.

Reply:
Should the `<=` there be `<`?

Reply:
For Ord it is `<`. Conceptually it is more `<=`. So I documented `<` here.

On wasm32 `Pointer` and `Bytes(4)` would be equivalent, but still you would consider Pointer as the (potentially) larger one, for Bytes(8) it is the other way round.